File: leetcode/368-largest-divisible-subset.py
class Solution(object):
    def largestDivisibleSubset(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        if 0 == n:
            return []

        nums.sort()

        maxcnt, maxpos = 1, 0
        v = [[nums[i]] for i in range(n)]
        for i in range(1, n):
            cnt, pos = 0, -1
            for j in range(i):
                if 0 == nums[i] % nums[j] and len(v[j]) > cnt:
                    cnt, pos = len(v[j]), j
            if pos >= 0:
                v[i] += v[pos]
                if len(v[i]) > maxcnt:
                    maxcnt, maxpos = len(v[i]), i

        return v[maxpos]

File: leetcode/368-largest-divisible-subset.cpp
class Solution
{
 public:
  vector<int> largestDivisibleSubset(vector<int>& nums)
  {
    int n = nums.size();
    if (0 == n) return {};

    sort(nums.begin(), nums.end());
    vector<vector<int>> v(n, vector<int>());

    int maxcnt = 1, maxpos = 0;
    v[0].push_back(nums[0]);
    for (int i = 1; i < n; ++i) {
      int pos = -1;
      for (int j = 0, cnt = 0; j < i; ++j) {
        if (0 == nums[i] % nums[j] && v[j].size() > cnt)
          cnt = v[j].size(), pos = j;
      }
      if (pos >= 0) {
        v[i] = v[pos];
        if (v[i].size() >= maxcnt)
          maxcnt = v[i].size() + 1, maxpos = i;
      }
      v[i].push_back(nums[i]);
    }

    return v[maxpos];
  }
};
